experimental/dds/tree2/src/shared-tree/undoRedoHandler.ts

```diff
@@ -0,0 +1,65 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { IRevertible, UndoRedoStackManager } from "@fluidframework/undo-redo";
+import { UndoRedoManagerCommitType } from "../core";
+import { ISharedTreeView } from ".";
+
+/**
+ * A shared tree view undo redo handler that will add all local undoable tree changes to the provided
+ * undo redo stack manager.
+ *
+ * @alpha
+ */
+export class SharedTreeViewUndoRedoHandler {
+	private detach?: () => void;
+
+	public constructor(private readonly stackManager: UndoRedoStackManager) {}
+
+	public attachTree(tree: ISharedTreeView) {
+		this.detach = tree.events.on("undoable", this.treeDeltaHandler);
+	}
+	public detachTree(tree: ISharedTreeView) {
```

Comment:
It sort of looks like this interface was being designed to support the attachment of multiple views, but then didn't. 

If only one view is supported, then I would suggest just having the view be passed into the constructor, removing `attachTree` and renaming `detachTree` to the more canonical `dispose`. At that point, there is only one method on the class that, when called, destroys the class... so I think the whole thing can collapse into a single function that mirrors the eventing registration pattern:
```
/** @returns a function that un-attaches **/
function attachView(view: SharedTreeView): () => void
```

If we want to support multiple views being attached, then we'll need to store a map of detach functions rather than just one. There is also the question of what happens if you attach the same view twice (do nothing? fail?), and if you try to detach a view that is not already attached (do nothing? fail?), and whether or not the many views share a single `UndoRedoStackManager` or each get their own.

Reply:
Oh very good point. It is kind of unfortunate we have the react style eventing here. I would argue we should support multiple views because every other undo redo handler supports multiple DDSes. The map approach isn't great but I think it's straightforward enough.